idxmgmt/ilm/config.go

```diff
@@ -18,34 +18,117 @@
 package ilm
 
 import (
-	"github.com/elastic/beats/libbeat/common/fmtstr"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
 	libilm "github.com/elastic/beats/libbeat/idxmgmt/ilm"
 )
 
-const pattern = "000001"
-
-// Config for ILM supporter
+//Config holds information about ILM mode, overwriting and policies
 type Config struct {
-	Mode       libilm.Mode               `config:"enabled"`
-	PolicyName *fmtstr.EventFormatString `config:"policy_name"`
-	AliasName  *fmtstr.EventFormatString `config:"alias_name"`
-	Event      string                    `config:"event"`
+	Mode          libilm.Mode `config:"enabled"`
+	Overwrite     bool        `config:"overwrite"`
+	RequirePolicy bool        `config:"require_policy"`
+	Policies      []Policy
 }
 
-// Enabled indicates whether or not ILM should be enabled
+//Policy binds together an ILM policy's name and body with an event type
+type Policy struct {
+	Policy    map[string]interface{}
+	EventType string
+	Name      string
+}
+
+//Enabled indicates whether or not ILM should be enabled
 func (c *Config) Enabled() bool {
 	return c.Mode != libilm.ModeDisabled
 }
 
-// ModeString stringifies enabled mode
-// This is a workaround as strings should be generated in libbeat.
-func ModeString(m libilm.Mode) string {
-	switch m {
-	case libilm.ModeAuto:
-		return "auto"
-	case libilm.ModeEnabled:
-		return "true"
-	default:
-		return "false"
+//NewConfig returns an ILM config, where default configuration is merged with user configuration
+func NewConfig(ucfg *common.Config) (Config, error) {
+	policies := policyPool()
+	policyMappings := policyMapping()
+	c := defaultConfig()
+	if ucfg != nil {
+		if err := ucfg.Unpack(&c); err != nil {
+			return Config{}, err
+		}
+
+		var cfg config
+		if err := ucfg.Unpack(&cfg); err != nil {
+			return Config{}, err
+		}
+		// create a collection of default and configured policies
+		for name, policy := range cfg.Policies {
+			policies[name] = policy
+		}
+		//update policy name per event according to configuration
+		for _, entry := range cfg.Mapping {
+			if _, ok := policyMappings[entry.Event]; !ok {
+				return c, errors.Errorf("event_type '%s' not supported for ILM setup", entry.Event)
+			}
+			policyMappings[entry.Event] = entry.Policy
+		}
+	}
+
+	for event, policyName := range policyMappings {
+		policy, ok := policies[policyName]
+		if !ok {
+			if c.RequirePolicy {
+				return Config{}, errors.Errorf("policy '%s' not configured for ILM setup", policyName)
+			}
+			policy = nil
```

Comment:
isn't this redundant?

Reply:
Why would that be redundant?

If the default configuration configures a policy for event_type x, but the user chooses to configure it referring to a policy that is not loaded via apm server, we have to ensure that no policy is loaded for this event type.

Reply:
maybe i'm missing something, but if a map lookup fails it returns (`nil`, `false`), so in the `if !ok` branch it is not needed to do `policy = nil` because it is `nil` already?